lib/instrumentation/langchain/tools.js

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2024 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+const { LangChainTool } = require('../../llm-events/langchain')
+const { mergeMetadata, mergeTags, recordEvent, shouldSkipInstrumentation } = require('./common')
+const {
+  AI: { LANGCHAIN }
+} = require('../../metrics/names')
+const { langchainRunId } = require('../../symbols')
+const { DESTINATIONS } = require('../../config/attribute-filter')
+
+module.exports = function initialize(shim, tools) {
+  const { agent, pkgVersion } = shim
+
+  if (shouldSkipInstrumentation(agent.config)) {
+    shim.logger.debug(
+      'langchain instrumentation is disabled.  To enable set `config.ai_monitoring.enabled` to true'
+    )
+    return
+  }
+
+  shim.record(tools.StructuredTool.prototype, 'call', function wrapCall(shim, call, fnName, args) {
+    const { name, metadata: instanceMeta, description, tags: instanceTags } = this
+    if (!shim.isWrapped(this, '_call')) {
+      shim.wrap(this, '_call', wrapUnderCall)
+    }
+
+    const [request, params] = args
+    const { metadata: paramsMeta, tags: paramsTags } = params || {}
+    return {
```

Comment:
Now that #2004 is merged, we should start utilizing the classes it added.

Reply:
so in shim.record is calls the [RecorderSpec](https://github.com/newrelic/node-newrelic/blob/main/lib/shim/shim.js#L679). do we want to back out all the direct constructions as right now it's doing it twice in some cases.

Reply:
The whole idea was that we'd be able to see at the creation point:

1. It is a spec object being created
2. What sort of spec is being created

If we remove the construction in the line you linked we may lose validation that the object works as desired. But it'd solve the double processing.

I think this situation is a result of the scattered object literals across the codebase. We don't know where things are being created, so we can't really go and replace all of them in one fell swoop.

Reply:
ok so it sounds like we need a ticket to remove constructing specs in shim and move it to the relevant instrumentation. in the meantime we will have places that do double construction, correct?

Reply:
Yes, I believe that is an accurate summation of the current state.

Reply:
Logged this follow up ticket #2022

Reply:
🤔 I just had a brainwave. PR incoming.